draft-pda-protocol.md

```diff
@@ -80,282 +79,294 @@ TODO: writeme
+  PPM system. In order to provide reasonable levels of privacy, there
+  must be a large number of clients.
 
-A common PPM task that can't be solved efficiently with Prio is the
-`t`-*heavy-hitters* problem {{BBCp21}}. In this setting, each user is in
-possession of a single `n`-bit string, and the goal is to compute the compute
-the set of strings that occur at least `t` times. One reason that Prio doesn't
-apply to this problem is that the proof generated by the client would be huge.
+Aggregator:
+: An endpoint which receives report shares. Each aggregator works with the
+  other aggregators to compute the final aggregate. This protocol defines
+  two types of aggregators: Leaders and Helpers. For each measurement,
+  there is a single leader and helper.
+
+Leader:
+: The leader is responsible for coordinating the protocol. It receives
+  the encrypted shares, distributes them to the helpers, and orchestrates
+  the process of computing the final measurement as requested by
+  the collector.
+
+Helper:
+: Helpers are responsible for executing the protocol as instructed
+  by the leader. The protocol is designed so that helpers can be relatively
+  lightweight, with most of the state held at the leader.
+{:br}
+
+
+The basic unit of PPM is the "task" which represents
+a single measurement (though potentially taken over multiple
+time wndows). The definition of a task includes the
+following parameters:
+
+* The values to be measured;
+* The statistic to be computed (e.g., sum, mean, etc.); 
+* The set of aggregators and necessary cryptographic keying material to use; and
+* The PPM scheme to use. This is to some extent dictated by the previous
+  choices.
+* The minimum "batch size" of reports which canbe aggregated.
+
+These parameters are distributed out of band to the clients and to
+the aggregators. Each task is identified by a unique 32-byte ID.
+
+During the duration of the measurement, each client records its own
+value(s), packages them up into a report, and sends them to the leader.
+Each share is separately encrypted for each aggregator so that even
+though they pass through the leader, it is unable to see or modify
+them. Depending on the measurement, the client may only send one
+report or may send many reports over time.
+
+The leader distributes the shares to the helpers and orchestrates
+the process of verifying them (see {{validating-inputs}})
+and assembling them into a final measurement for the collector.
+Depending on the PPM scheme, it may be possible to incrementally
+process each report as it comes in, or may be necesary to wait
+until all the required shares are received. 
+
+This protocol is compatible with both a "push" deployment mode in which the 
+leader computes the measurements at predetermined points and sends the results
+to the collector and a "pull" mode in which the collector tells
+the leader when to perform the aggregation and return the results.
```

Comment:
Oh, I see. By "push" you mean there might be some alternative to an HTTP request for gathering a CollectResp? Remember that, in general, thee collector needs to convey CollectReq to the aggregators somehow. (For instance, in Hits, the output can't be produced until the collector hands the aggregators the set of candidate prefixes.)

If that's what you mean, an open issue sounds fine.

Reply:
Wood and I spent a while discussing this on IM and it seems
to me that there are a few pieces here:

- Incremental computation
- Proactively producing results

To take Prio as our example, it's quite possible for the leader to
just send every report to the helpers as it comes in, then issue
OutputShareReq when the batch window + jitter window expires and then
just store the data for CollectReq.  It seems like that has some
obvious advantages, though note that it's soft state and is invisible
to the collector except for timing.

Moving to Hits, the situation seems a bit more complicated.  As I
understand it, it's not possible to collect all hits above a given
threshold and have the output concealed from the leader [0]. It
might, however, be possible to verify the proofs first, or at
least be theoretically possible even if Hits doesn't allow it
now. Again, this would be soft state.


Conversely, I think it's clear we're going to need some way for the
collector to set off a long running computation and get the results
when its done. This could either be via a long poll/ websockets or
some sort of notification hook. It's possible that the leader could
use this mechanism whatever it is to proactively inform the collector
of results (i.e., publish/subscribe), though maybe that's premature
here.

IOW I think it's important that we don't say that this protocol
requires that all AggregateReq/OutputShare operations are triggered
by a CollectReq. I think it's less important that there be some way
for it to automatically do that and deliver the results. I can remove
that and/or add an OPEN ISSUE.


[0] Though presumably the pattern of queries allows the
leader to infer which strings are of interest, right?

Reply:
> To take Prio as our example, it's quite possible for the leader to
> just send every report to the helpers as it comes in, then issue
> OutputShareReq when the batch window + jitter window expires and then
> just store the data for CollectReq. It seems like that has some
> obvious advantages, though note that it's soft state and is invisible
> to the collector except for timing.

Yup!

> Moving to Hits, the situation seems a bit more complicated. As I
> understand it, it's not possible to collect all hits above a given
> threshold and have the output concealed from the leader [0]. It
> might, however, be possible to verify the proofs first, or at
> least be theoretically possible even if Hits doesn't allow it
> now. Again, this would be soft state.

This isn't possible for the current Hits protocol, sadly. The input validation protocol involves first evaluating the IDPF on the set of candidate prefixes, then running the "secure sketching" protocol of Section 4 to make sure that at most of one of these prefixes evaluates to 1 (and the others evaluate to 0). That means that input validation has to wait until the candidate prefixes are disseminated (from the collector to the aggregators). Maybe there's an alternative protocol for which input validation doesn't depend on the prefixes, but we don't have it yet.

> Conversely, I think it's clear we're going to need some way for the
> collector to set off a long running computation and get the results
> when its done. This could either be via a long poll/ websockets or
> some sort of notification hook. It's possible that the leader could
> use this mechanism whatever it is to proactively inform the collector
> of results (i.e., publish/subscribe), though maybe that's premature
> here.

Agreed. it's definitely going to be the case that the time between a collect request is issued and the response is generated could be a matter of hours. Though I wonder if it's as simple as responding to as collect request by saying "call back in two hours, I should have your answer ready by then"? Do we really need something fancier than this?

> IOW I think it's important that we don't say that this protocol
> requires that all AggregateReq/OutputShare operations are triggered
> by a CollectReq.

If we want to implement Hits, then we have to accept that, in general, input validation and aggregation depends on parameters disseminated by the collector. Saying otherwise means trying to fit Hits into a box it doesn't fit in.

Of course, we could consider changing the shape of the box. In particular, we might remove the encryption from the output share request and have the leader combine the output shares to get the candidate prefixes and initiate the next round itself. This way the collector would only be involved at the very end, once all `n` collect rounds (i.e., rounds of input validation and aggregation) are completed. (Is this what you're suggesting?)

It's not clear to me that this results in any significant speed-up. It seems like the main advantage of this approach is that it eliminates the interaction with the collector at each round. The downside is that it rules out uses cases brought up @csharrison in which the collector wants the opportunity to "prune" the search space at each round based on knowledge it has of the input space. Wherever this optimization is possible, it is sure to bring a significant performance boost.

> I think it's less important that there be some way
> for it to automatically do that and deliver the results. I can remove
> that and/or add an OPEN ISSUE.

By "less" do you mean "more" here?

> [0] Though presumably the pattern of queries allows the
> leader to infer which strings are of interest, right?

Right, there's no way to avoid this with the current Hits protocol.

By the way, this conversation raises an interesting question. Right now we have just two PPM protocols, one in which input validation and aggregation are triggered by parameters that need to be disseminated to the aggregators (Hits), and another in which these steps can proceed immediately (Prio). The question is: _For other PPM protocols, which of these scenario is more likely?_ I don't think we have an answer right now.